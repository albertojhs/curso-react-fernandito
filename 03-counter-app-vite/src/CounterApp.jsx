import { useState } from 'react';
import PropTypes from "prop-types";


export const CounterApp = ({ value }) => {
    
    /* Declaración de constantes */
    const [ counter, setCounter ] = useState( value );


    

    /*Declaración de funciones */
    const handleAdd = () => { 
        setCounter( counter + 1 ); 
    }

    const handleSub = () => {
        setCounter( counter - 1 );
    }

    const handleReset = () => {
        setCounter( value );
    }




    /*Return para el render */
    return (
        <>
            <h2>CounterApp</h2>
            <h3> { counter } </h3>
            <button onClick={ handleAdd }>
                +1
            </button>
            <button onClick={ handleSub }>
                -1
            </button>
            <button aria-label="btn-reset" onClick={ handleReset }>
                reset
            </button>
        </>
    );
} 



/*Tipado, restricciones, etc del props */
CounterApp.propTypes = {
    value : PropTypes.number.isRequired
}