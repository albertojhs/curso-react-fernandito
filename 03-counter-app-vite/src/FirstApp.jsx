import { Fragment } from 'react'
import PropTypes from 'prop-types';

//Esta variable aquí no está declarada en el scope global, sigue estando encapsulada en el componente
//pero se recomienda que se coloquen las constantes y variables que no tengan relación directa con las
//funsiones fuera de las mismas.

// const myObj = {
//     name : 'Diego Teolotitla',
//     edad : 78
// }

export const FirstApp = ( { title, subtitle, number } ) => {


    //Simepre se debe colocar un nodo padre en el return


    // return (
    //     <Fragment>
    //         <h1>First App</h1>
    //         <p>Soy solo un texto, sit amet consectetur adipisicing elit. Itaque, inventore! Fugit deleniti repellat eligendi maiores nam doloremque corrupti, ipsum quisquam, distinctio culpa nihil iure sunt facere quae? Velit, ipsam autem!</p>
    //         <p>Lorem ipsum dolor sit, amet consectetur adipisicing elit. Earum, assumenda repellendus. Dolor, cupiditate ducimus fugit quo voluptates, dolorum repellendus ratione nulla nisi consequuntur inventore perspiciatis a, vero iusto ad. Id.
    //         Officia ea a nemo eligendi debitis eveniet animi quia cum laboriosam, recusandae laudantium alias unde laborum culpa molestias iusto ipsum autem incidunt nihil quo! Blanditiis optio ducimus iste nostrum officiis.
    //         Similique dicta mollitia provident sapiente quas fugiat vitae sequi fuga esse eum dolorum hic earum sunt deleniti, rem commodi ab ipsam repudiandae. Molestiae tenetur beatae repellendus, in amet expedita perferendis!
    //         Excepturi maxime error aspernatur ea facilis magnam quis magni quos, sequi iusto commodi adipisci, placeat deserunt molestiae nisi tempora consequuntur atque at et similique inventore autem quas dicta. Officiis, illum.</p>
    //     </Fragment>
    // )

    // <> Es equivalente a lo de arriba solo que no es necesario importar el fragment

    return (
        <>
            <h1 data-testid="test-title">{ title }</h1>
            <h2>{ subtitle }</h2>
            <h2>{ number }</h2>
            <p>Sit amet consectetur adipisicing elit. Itaque, inventore! Fugit deleniti repellat eligendi maiores nam doloremque corrupti, ipsum quisquam, distinctio culpa nihil iure sunt facere quae? Velit, ipsam autem!</p>
            <p>Lorem ipsum dolor sit, amet consectetur adipisicing elit. Earum, assumenda repellendus. Dolor, cupiditate ducimus fugit quo voluptates, dolorum repellendus ratione nulla nisi consequuntur inventore perspiciatis a, vero iusto ad. Id.
            Officia ea a nemo eligendi debitis eveniet animi quia cum laboriosam, recusandae laudantium alias unde laborum culpa molestias iusto ipsum autem incidunt nihil quo! Blanditiis optio ducimus iste nostrum officiis.
            Similique dicta mollitia provident sapiente quas fugiat vitae sequi fuga esse eum dolorum hic earum sunt deleniti, rem commodi ab ipsam repudiandae. Molestiae tenetur beatae repellendus, in amet expedita perferendis!
            Excepturi maxime error aspernatur ea facilis magnam quis magni quos, sequi iusto commodi adipisci, placeat deserunt molestiae nisi tempora consequuntur atque at et similique inventore autem quas dicta. Officiis, illum.</p>
            {/* <code>{ JSON.stringify(myObj) }</code> */}
        </>
    )
};


FirstApp.propTypes = {
    subtitle : PropTypes.string.isRequired,
    title    : PropTypes.string.isRequired,
    number   : PropTypes.number
}


// FirstApp.defaultProps = {
//     name     : 'Alberto Hernández',
//     subtitle : 'No hay subtitlo',
//     title    : 'No hay título'
// }