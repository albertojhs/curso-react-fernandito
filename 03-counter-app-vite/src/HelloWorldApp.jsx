export const App = () => (<h1>Hola mundo de React!</h1>);