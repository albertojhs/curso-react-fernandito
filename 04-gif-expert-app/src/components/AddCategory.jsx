import { useState } from 'react';
import { PropTypes } from 'prop-types';

export const AddCategory = ({onAddCategory}) => {

    const [inputValue, setInputValue] = useState('');

    const onInputChange = ( {target} ) => {
        setInputValue( target.value );
    }

    const onSubmit = (event) => {
        event.preventDefault();
        if( inputValue.trim().length < 1 ) return;
        onAddCategory( inputValue.trim() );
        setInputValue( '' );

        //Si usaramos setCategories en su lugar habría que pasar un callback en lugar de un string
        //Para esto tendríamos que mandar el setCategories en lugar de onAddCategory en el 
        //componente de GifExpertApp, esto no es muy recomendable ya que quien use GifExpertApp
        //debe de saber que se trata de una función de state y hace las cosas un poco difíciles de
        //usar
        // props.setCategories( categories => [ inputValue, ...categories] );
        
    }


    return (
        // <form onSubmit={ (event) => onSubmit(event) }>
        <form onSubmit={ onSubmit } aria-label="form">
            <input 
                type="text" 
                placeholder="Buscar gifs"
                value={ inputValue }
              //   onChange={ (event) => onInputChange(event) }
                onChange={ onInputChange }
            />
        </form>
    )
}

AddCategory.propTypes = {
    onAddCategory:  PropTypes.func.isRequired
}
