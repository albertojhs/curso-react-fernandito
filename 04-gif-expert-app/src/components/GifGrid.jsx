
// import { useState, useEffect } from 'react';
// import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';
import { GifItem } from './GifItem';

export const GifGrid = ({ category }) => {


  //Este juc personalizado equivale a lo de abajo más la lógica del 
  //isloading

  const { images, isLoading } = useFetchGifs( category ); 

  // const [images, setImages] = useState([]);

  // //Si se dejan las dependencias vacías significa que ese hook se va a disparar solo la primera vez
  // useEffect( () => {
  //   getGifs( category )
  //     .then( (imgs) => {
  //       setImages(imgs);
  //     } );
  // }, [] );


  return (
    <>
        <h3>{ category }</h3>
        {/* {
          isLoading ? 
          ( <h2>Cargando...</h2> ) :
          null
        } */}
        {
          isLoading && ( <h2>Cargando...</h2> )
        }
        <div className='card-grid'>
          { images.map( (img) => (
            //Estamos esparciendo las propiedades para que puedan ser desestructuradas en 
            //GifItem ( {...img} )
            <GifItem key={ img.id } { ...img }/>
          ))}
        </div>
    </>
  )
}

GifGrid.propTypes = {
  category : PropTypes.string.isRequired
}