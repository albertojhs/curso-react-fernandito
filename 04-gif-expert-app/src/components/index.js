// Archivo de barril del chavo del 8
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';