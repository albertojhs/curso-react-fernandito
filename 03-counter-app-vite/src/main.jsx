import React from 'react';
import ReactDOM from 'react-dom/client';

import './styles.css';
import { CounterApp } from './CounterApp';


ReactDOM.createRoot( document.getElementById('root') ).render(
    <CounterApp value={27}/>
);