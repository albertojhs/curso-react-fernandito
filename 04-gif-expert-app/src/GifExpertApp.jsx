import { useState } from 'react';
import { AddCategory, GifGrid } from "./components"; //Por defecto busca el index de la carpeta que es mi archivo de barril


export const GifExpertApp = () => {

  const [categories, setCategories] = useState([ 'UFO' ]);

  const onAddCategory = ( newCategory ) => {
    setCategories( [ newCategory, ...categories.filter( cat => cat.toLowerCase() !== newCategory.toLowerCase()) ] );
  }

  return (
    <>
        <h1>GifExpertApp</h1>
        <AddCategory onAddCategory={ onAddCategory } />
        { categories.map( category => ( 
            <GifGrid 
              key={ category } 
              category={ category }/>
          ))
        }
    </>
  )
}
