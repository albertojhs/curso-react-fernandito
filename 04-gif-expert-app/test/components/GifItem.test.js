import { render, screen } from '@testing-library/react';

import { GifItem } from "../../src/components";

describe('Pruebas de GifItem', () => { 

    const titulo = 'Texto alternativo para el gif';
    const url = 'https://www.pagina-de-gifs.com/';
    
    test('GifItem debe de validar el snapshot', () => {
        const { container } = render( <GifItem title={ titulo } url={ url }/> );
        expect( container ).toMatchSnapshot();
    });

    test('Debe de mostrar la imágen con el url y el alt indicado', () => {
        render( <GifItem title={ titulo } url={ url }/> );
        // screen.debug();
        
        // expect( screen.getByRole('img').src ).toBe( url );
        // expect( screen.getByRole('img').alt ).toBe( titulo );
        
        const { src, alt } = screen.getByRole('img');
        expect( [ src, alt ] ).toStrictEqual( [ url, alt ] );
    });
    
    test('Debe de mostrar el título  en el componente', () => {
        render( <GifItem title={ titulo } url={ url }/> );
        expect( screen.getByText( titulo ) ).toBeTruthy();
    });
 })