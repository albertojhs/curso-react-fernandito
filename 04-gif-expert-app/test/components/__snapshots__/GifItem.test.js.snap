// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de GifItem GifItem debe de validar el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Texto alternativo para el gif"
      src="https://www.pagina-de-gifs.com/"
    />
    <p>
      Texto alternativo para el gif
    </p>
  </div>
</div>
`;
